feat: add support for `vm.compileFunction`

SimenB opened the pull request:
Id like to move Jest away from the module wrapper, similar to what Node core did here: https://github.com/nodejs/node/pull/21573. Jest currently uses its own module wrapper, puts it in a `Script`, then executes it using `runVMScript`. However, if we use `compileFunction` instead, we don't need to create the wrapper.

Note that this function was added in `node@10.10`, so I've added a guard that'll throw on node 8 (or some minor of node that don't have the function). I'd prefer this to land in the current version of JSDOM without bumping the major, as otherwise we'd be unable to use this in Jest until sometime next year. I probably need to add a guard in the tests for this case, but I'd like to get some feedback on this first approach first.

domenic commented:
We won't be releasing any more majors of jsdom with the current version, so I'm sorry that won't be possible. Perhaps we could bump the minimum required Node.js version to 10 at the same time.

I'm not sure in general whether we want to add this complexity to jsdom to avoid "a wrapper", but I'm having a hard time getting a sense of what exactly the wrapper is. Can you try to explain the benefit to users in more detail? The example code you've added isn't very compelling; compileVMFunction and runVMScript seem very redundant. Perhaps we should remove runVMScript?

SimenB commented:
Yeah, the OP was pretty parse in this case and focused more on a solution to a problem I'm facing than why it should be in JSDOM. Sorry about that.

---

>We won't be releasing any more versions of jsdom with the current (v15) major, so I'm sorry that won't be possible. Perhaps we could bump the minimum required Node.js version to 10 at the same time.

Ok, that's fine 🙂 If you're going for a major soonish, dropping node 8 should be no issue as it's EOL 2019-12-31 anyways (https://github.com/nodejs/Release/blob/master/README.md). That neatly solves the issue of checking for support at the same time.

>Can you try to explain the benefit to users in more detail?

Sure! I'll try to, by taking a step back. Correct me if I'm wrong, but the main use case for `runVMScript` is to be able to pass some JS, have it evaluated within the JSDOM context, and pass back the result of this evaluation. Jest currently uses this by creating a `vm.Script` with e.g. this content

```js
({
  evalResult: function(arg1, arg2) {
    // do something in here
  }
})
```

After running the script, Jest then does `scriptResult.evalResult.call(null, passAsArg1, passAsArg2)` allowing it to inject `require`, `module` etc. from the outside. This has some issues when trying to map source code back, especially for code coverage.

There are more details about the general problem with the wrapper here: https://github.com/nodejs/node/issues/17396

Using `vm.compileFunction` instead allows us to get a function back directly, without the need to create a wrapper.

>compileVMFunction and runVMScript seem very redundant. Perhaps we should remove runVMScript?

Yup, totally! The advantage (in my mind) of `runVMScript` is that a `vm.Script` can take the content of a whole JS file, not just the body of a function. However, if your use case is "this whole file", you should probably use script tags, `eval` or similar, like documented in the readme today. The use case of "evaluate a function in the context of JSDOM and invoke that function from the outside" is better covered by `compileVMFunction`, like I talked about above. However, there might be use cases for `runVMScript` that I don't know about?

thymikee commented on lib/api.js:
```diff
@@ -80,6 +80,27 @@ class JSDOM {
+    options = options || {};
+
+    if (options.parsingContext) {
+      throw new TypeError("You cannot pass a parseing context to compileFunction");
```
```suggestion
      throw new TypeError("You cannot pass a parsing context to compileFunction");
```

SimenB commented:
@domenic thoughts? I'm happy to remove `runVMScript` at the same time

domenic commented:
Sorry for the delay!

Let's try removing runVMScript, indeed. When preparing the next release, I'll add a note to the changelog and/or readme that if compileVMFunction doesn't fit folks' use case, then we can add it back, but for now I buy your reasoning on compileVMFunction being generally more useful, and I'd like to keep the more-esoteric methods to a minimum if that works for the community.

Feel free to remove Node v8 from the matrix (bumping the node-canvas test suite to v10), and change the package.json engines field, to get the CI to pass.

ExE-Boss commented on test/api/methods.js:
```diff
@@ -191,6 +191,83 @@ describe("API: JSDOM class's methods", () => {
+    });
+
+    it("should work when runScripts is set to \"dangerously\"", () => {
+      const dom = new JSDOM(``, { runScripts: "outside-only" });
```
```suggestion
      const dom = new JSDOM(``, { runScripts: "dangerously" });
```

SimenB commented:
Thanks @domenic, I've updated this now. In addition to the changes you mentioned, I also added node 12 to the testing matrix, as it's the current LTS version.

SimenB commented on README.md:
```diff
@@ -309,32 +309,32 @@ console.log(dom.nodeLocation(imgEl));    // { startOffset: 13, endOffset: 32 }
 Note that this feature only works if you have set the `includeNodeLocations` option; node locations are off by default for performance reasons.
 
-### Running vm-created scripts with `runVMScript(script[, options])`
+### Compiling a function with `compileVMFunction(code[, params[, options]])`
```
I'm not sure about the notation for multiple optional arguments

SimenB commented:
Actually thinking about this, I wonder what it'd mean for modules? Those can't be wrapped in functions... We'll probably (when the time comes) inject Jest's "globals" through a loader when that API is ready, but I suppose that'll be closer to the current `vm.Script` approach than this. Some discussion here, but it's basically "we don't know yet": https://github.com/nodejs/node/issues/27387. Currently `vm.Script` doesn't work with ESM anyways, so it doesn't matter much, but something to keep in mind

domenic commented:
Hmm. Maybe we should take a step back, and instead just directly expose something like `dom.vmContext` (which returns `this[window]`)? And have the README give a few examples of how to use it, and have the tests test a couple main-line use cases like `compileFunction` and `runScript`? It would return `null` (or `undefined`?) if `runScripts` was not set.

That should give maximum flexibility for integrators like Jest, including future things like modules.

It's a bit of a foot-gun if people start, for whatever reason, using `dom.vmContext` instead of `dom.window`, but hopefully the name would discourage them from doing that.

SimenB commented:
Sorry about the slow response, currently semi offline due to holidays (or, I try to be with varying luck 😛).

I agree a lower level API exposing the raw context directly makes sense, that way jsdom can be as flexible as possible for the advanced use cases.

I'd go with a `getInternalVmContext()` or something as its name to make it very explicit it's not an API you'd use normally. I'm happy to change this PR to be that if you want, although I won't get to it until the week after new years.
